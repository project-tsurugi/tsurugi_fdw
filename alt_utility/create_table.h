#ifndef CREATE_TABLE_
#define CREATE_TABLE_

#ifdef __cplusplus
extern "C" {

#endif

bool create_table(CreateStmt *stmt);

#ifdef __cplusplus
}
#endif

#endif // CREATE_TABLE_
