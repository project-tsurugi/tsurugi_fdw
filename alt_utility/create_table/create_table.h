/*
 * Copyright 2019-2020 tsurugi project.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 *	@file	  table_metadata.h
 *	@brief  TABLE metadata operations.
 */
#pragma once

#include <boost/property_tree/ptree.hpp>
#include "command/create_command.h"

class CreateTable : public CreateCommand {
 public:
  CreateTable(CreateStmt* create_stmt) : CreateCommand(create_stmt) {}

  /**
   *  @brief  Check if given syntax supported or not by Tsurugi
   *  @return true if supported
   *  @return false otherwise.
   */
  virtual bool validate_syntax();

  /**
   *  @brief  Check if given syntax supported or not by Tsurugi
   *  @return true if supported
   *  @return false otherwise.
   */
  virtual bool validate_data_type();

  /**
   *  @brief  Create table metadata from query tree.
   *  @return true if supported
   *  @return false otherwise.
   */
  virtual bool generate_metadata(boost::property_tree::ptree& metadata);

 private:
  bool create_column_metadata(ColumnDef* colDef, 
                              int64_t ordinal_position, 
                              boost::property_tree::ptree& column);
  bool put_data_lengths(List* typmods, boost::property_tree::ptree& datalengths);
};
