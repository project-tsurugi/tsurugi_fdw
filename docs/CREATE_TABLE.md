# PostgreSQLからのテーブル定義
2020.04.01 NEC

## 概要

* 本文書では、V1におけるPostgreSQLからNEDO DBにテーブルを定義する手順について説明する。
* 本文書はpsqlまたはPL/pgSQLからテーブルを定義することを想定している。

## 事前準備

psqlから以下の手順でFDWを利用可能にする。。

1. FDWのインストール
	* CREATE EXTENSION文を実行する
		```sql
		CREATE EXTENSION ogawayama_fdw;
		```
	* メタコマンド(\dew)で確認する
		```
		postgres=# \dew
                        List of foreign-data wrappers
             Name      |  Owner   |        Handler        | Validator
		---------------+----------+-----------------------+-----------
 		 ogawayama_fdw | postgres | ogawayama_fdw_handler | -
		```

1. 外部サーバの定義
	* CREATE SERVER文を実行する
		```sql
		CREATE SERVER ogawayama FOREIGN DATA WRAPPER ogawayama_fdw;
		```
	* メタコマンド(\des)で確認する
		```
		postgres=# \des
                    List of foreign servers
   		   Name    |  Owner   | Foreign-data wrapper
		-----------+----------+----------------------
 		 ogawayama | postgres | ogawayama_fdw
		```

1. TABLESPACEの定義
	* TABLESPACE用のディレクトリを作成する
		```
		$ mkdir -p (PostgreSQLインストール先)/data/tsurugi
		```
	* CREATE TABLESPACE文を実行する
		```sql
		CREATE TABLESPACE tsurugi LOCATION '(PostgreSQLインストール先)/data/tsurugi';
		```
	* メタコマンド(\db)で確認する
		```
		k-postgres=# \db
                   List of tablespaces
			Name    |  Owner   |             Location
		------------+----------+-----------------------------------
		 pg_default | postgres |
		 pg_global  | postgres |
		 tsurugi    | postgres | /home/postgres/local/pgsql/data/tsurugi
		```

## テーブル定義手順

1. テーブルの定義
	* CREATE TABLE文を実行する
		* "TABLESPACE tsurugi"を付加する
		* **必ず主キー(PRIMARY KEY)を指定する**
			```sql
			CREATE TABLE table1 (column1 INTEGER NOT NULL PRIMARY KEY) TABLESPACE tsurugi;
			```
	* PostgreSQLには指定したテーブル名に"_dummy"が付加されたテーブル名で定義される
		* e.g. "table1_dummy"

1. 外部テーブルの定義
	* CREATE FOREIGN TABLE文を実行する
		* 先に定義したテーブルと同じテーブル名、テーブルスキーマで定義する
		* **ただし、主キーは指定しない**
		* 先に定義した外部サーバを指定する
			```sql
			CREATE FOREIGN TABLE table1 (column1 INTEGER NOT NULL) SERVER ogawayama;
			```
	* 外部テーブルは指定したテーブル名で定義される
		* e.g. "table1"

1. DMLの実行
	* DML文は外部テーブルを指定して実行する
		```sql
		SELECT * FROM table1;
		INSERT INTO table1 VALUES (100);
		```

以上
