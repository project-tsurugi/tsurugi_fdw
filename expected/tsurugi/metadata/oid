tables=44
column=310
