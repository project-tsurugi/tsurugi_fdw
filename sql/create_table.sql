
/* DML */
SELECT * FROM t1_create_table;
INSERT INTO t1_create_table (c1) VALUES (1);
SELECT * FROM t1_create_table;

SELECT * FROM t2_create_table;
INSERT INTO t2_create_table (c1, c2, c3) VALUES (1, 100, 1.1);
INSERT INTO t2_create_table (c1, c2, c3) VALUES (2, 200, 2.2);
SELECT * FROM t2_create_table;

SELECT * FROM t1_create_table;
INSERT INTO t1_create_table (c1) VALUES (10);
INSERT INTO t1_create_table (c1) VALUES (100);
SELECT * FROM t1_create_table;

SELECT * FROM t2_create_table;
UPDATE t2_create_table SET c2 = c2+10;
UPDATE t2_create_table SET c3 = c3+1.1 WHERE c2 = 110;
SELECT * FROM t2_create_table;

SELECT * FROM t1_create_table;
DELETE FROM t1_create_table WHERE c1 = 10;
SELECT * FROM t1_create_table;

/* Fix tsurugi-issues#568 */
INSERT INTO public.t1_create_table (c1) VALUES (1000);
INSERT INTO PuBlIc.t1_create_table (c1) VALUES (2000);
INSERT INTO PUBLIC.t1_create_table (c1) VALUES (3000);
INSERT INTO "public"."t1_create_table" (c1) VALUES (4000);

SELECT * FROM public.t1_create_table;

UPDATE public.t1_create_table SET c1 = c1+100;
UPDATE PuBlIc.t1_create_table SET c1 = c1+100 WHERE c1 > 1000;
UPDATE public.t1_create_table SET c1 = c1+100 WHERE c1 > 2000;
UPDATE "public"."t1_create_table" SET c1 = c1+100 WHERE c1 > 3000;

SELECT * FROM "public"."t1_create_table";

DELETE FROM public.t1_create_table WHERE c1 = 200;
DELETE FROM "public"."t1_create_table" WHERE c1 > 1000;

SELECT * FROM Public.t1_create_table;

INSERT INTO "puBlIc"."t1_create_table" (c1) VALUES (999); -- error
UPDATE "PUBLIC"."t1_create_table" SET c1 = c1+100; -- error
DELETE FROM "Public"."t1_create_table" WHERE c1 > 1000; -- error
