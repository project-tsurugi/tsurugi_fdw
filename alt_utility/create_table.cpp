/*
 * Copyright 2019-2020 tsurugi project.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 *	@file	create_table.cpp
 *	@brief  Dispatch the create-table command to ogawayama.
 */

#include <string>
#include <string_view>
#include <regex>

#include "ogawayama/stub/api.h"
#include "stub_manager.h"
#include "manager/metadata/metadata.h"
#include "manager/metadata/datatypes.h"

using namespace boost::property_tree;
using namespace manager;
using namespace ogawayama;

#ifdef __cplusplus
extern "C" {
#endif
#include "postgres.h"
#include "nodes/parsenodes.h"
#ifdef __cplusplus
}
#endif

#include "tablecmds.h"
#include "create_table.h"

static std::string rewrite_query(std::string_view query_string);
static bool execute_create_table(std::string_view query_string);

/*
 *  @brief:
 */
bool create_table(CreateStmt *stmt)
{
    Assert(stmt != nullptr);

    TableCommands table_commands;
    bool success = table_commands.define_relation(stmt);
    if (!success)
    {
        elog(ERROR, "execute_create_table() failed.");
    }

    return success;
}

/*
 *  @brief:
 */
bool create_table(const char* query_string)
{
    Assert(query_string != nullptr);

    std::string query{query_string};

    bool success = execute_create_table(query);
    if (!success)
    {
        elog(ERROR, "execute_create_table() failed.");
    }

    return success;
}

/*
 *  @brief:
 */
bool execute_create_table(std::string_view query_string)
{
    bool ret_value = false;
    ERROR_CODE error = ERROR_CODE::UNKNOWN;

    const std::string rewrited_query = rewrite_query(query_string);

    // dispatch create_table query.
    stub::Transaction* transaction;
    error = StubManager::begin(&transaction);
    if (error != ERROR_CODE::OK)
    {
        std::cerr << "begin() failed." << std::endl;
        return ret_value;
    }

    elog(DEBUG2, "rewrited query string : \"%s\"", rewrited_query.c_str());
    error = transaction->execute_create_table(rewrited_query);
    if (error != ERROR_CODE::OK)
    {
        elog(ERROR, "transaction::execute_create_table() failed. (%d)", (int) error);
        return ret_value;
    }

    error = transaction->commit();
    if (error != ERROR_CODE::OK)
    {
        elog(ERROR, "transaction::commit() failed. (%d)", (int) error);
        return ret_value;
    }
    StubManager::end();

    ret_value = true;

    return ret_value;
}

/*
 *  @brief:
 */
static std::string rewrite_query(std::string_view query_string)
{
    std::string rewrited_query{query_string};

    std::unique_ptr<metadata::Metadata> datatypes{new metadata::DataTypes("NEDO DB")};

    metadata::ErrorCode error = datatypes->load();
    if (error != metadata::ErrorCode::OK) {
        std::cout << "DataTypes::load() error." << std::endl;
    }

    // trim a terminal semi-column.
    if (rewrited_query.back() == ';' ) {
        rewrited_query.pop_back();
    }

    ptree datatype;

    while ((error = datatypes->next(datatype)) == metadata::ErrorCode::OK) {
        boost::optional<std::string> pg_type_name =
            datatype.get_optional<std::string>(metadata::DataTypes::PG_DATA_TYPE_NAME);
        boost::optional<std::string> og_type_name =
            datatype.get_optional<std::string>(metadata::DataTypes::NAME);
        if (!pg_type_name.get().empty() && !og_type_name.get().empty()) {
            try {
                rewrited_query = std::regex_replace(
                    rewrited_query,
                    std::regex("(\\s)(" + pg_type_name.get() + ")([\\s,)])", std::regex_constants::icase),
                    "$1" + og_type_name.get() + "$3");
            } catch (std::regex_error e) {
                std::cout << "regex_replace() error. " << e.what() << std::endl;
            }
        }
    }

    try {
        rewrited_query = std::regex_replace(
            rewrited_query, std::regex("\\sTABLESPACE\\sTsurugi", std::regex_constants::icase), "");
    } catch (std::regex_error e) {
        std::cout << "regex_replace() error. " << e.what() << std::endl;
    }

    return rewrited_query;
}
