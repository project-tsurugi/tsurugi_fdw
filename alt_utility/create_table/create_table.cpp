/*
 * Copyright 2019-2020 tsurugi project.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 *	@file	  table_metadata.h
 *	@brief  TABLE metadata operations.
 */
#include "create_table.h"

#include <unordered_set>
#include <boost/property_tree/ptree.hpp>
#include <boost/optional.hpp>
#include "manager/metadata/metadata.h"
#include "manager/metadata/datatypes.h"
#include "manager/metadata/tables.h"
#include "manager/metadata/error_code.h"

#ifdef __cplusplus
extern "C" {
#endif
#include "postgres.h"
#include "catalog/pg_class.h"
#include "nodes/nodes.h"
#include "nodes/parsenodes.h"
#include "nodes/value.h"
#ifdef __cplusplus
}
#endif

using boost::property_tree::ptree;
using namespace manager;

/* base index of ordinal position metadata-manager manages */
const metadata::ObjectIdType ORDINAL_POSITION_BASE_INDEX = 1;

/**
 * @brief seek ordinal positions of primary keys in table or column constraints.
 * 
 * 
 * 
 */
std::vector<int64_t> get_primary_key(
    	const CreateStmt* create_stmt, 
		std::vector<metadata::Column>& columns)
{
	std::vector<int64_t> primary_keys;

	List* table_elts = create_stmt->tableElts;
	ListCell* listptr;
	foreach(listptr, table_elts) {
		// Seek primary-key in column constraints.
		Assert(IsA(lfirst(listptr), ColumnDef));
		ColumnDef* column_def = (ColumnDef *) lfirst(listptr);
		List* column_def_constraints = column_def->constraints;
		if (column_def_constraints != NIL) {
			ListCell* lc;
			foreach(lc, column_def_constraints) {
				Constraint* constr = (Constraint*) lfirst(lc);
				/* Get oridinal positions of column constraints' primary key columns */
				if (constr->contype == CONSTR_PRIMARY) {
					for (metadata::Column column : columns) {
						if (column.name == column_def->colname) {
							primary_keys.emplace_back(column.ordinal_position);
							break;
						}
					}
				}
			}
		}
	}

	return primary_keys;
}

/**
 *  @brief  Check if given syntax supported or not by Tsurugi
 *  @return true if supported
 *  @return false otherwise.
 */
bool CreateTable::validate_syntax() const
{
  assert(create_stmt() != nullptr);

  const CreateStmt* create_stmt = this->create_stmt();
  bool result{false};
  List* table_elts = create_stmt->tableElts;

  /* Check members of CreateStmt structure. */
  if (create_stmt->inhRelations != NIL) {
      ereport(ERROR,
              (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
                errmsg("Tsurugi does not support INHERITS clause")));
      return result;
  }

  if (create_stmt->partbound != nullptr) {
      ereport(ERROR,
              (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
                errmsg("Tsurugi does not support FOR VALUES clause")));
      return result;
  }

  if (create_stmt->partspec != nullptr) {
      ereport(ERROR,
              (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
                errmsg("Tsurugi does not support PARTITION BY clause")));
      return result;
  }

  if (create_stmt->ofTypename != nullptr) {
      ereport(ERROR,
              (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
                errmsg("Tsurugi does not support OF typename clause")));
      return result;
  }

  if (create_stmt->options != NIL) {
      ereport(ERROR,
              (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
                errmsg("Tsurugi does not support WITH clause")));
      return result;
  }

  if (create_stmt->oncommit != ONCOMMIT_NOOP) {
      ereport(ERROR,
              (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
                errmsg("Tsurugi does not support ON COMMIT clause")));
      return result;
  }

#if PG_VERSION_NUM >= 120000
  if (create_stmt->accessMethod != nullptr) {
      ereport(ERROR,
              (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
                errmsg("Tsurugi does not support USING clause")));
      return result;
  }
#endif

  /* for duplicate column name check */
  std::unordered_set<std::string> column_names;

  /* Check members of each column */
  ListCell* l;
  foreach(l, table_elts) {
    ColumnDef *column_def = (ColumnDef *)lfirst(l);
    List *column_def_constraints = column_def->constraints;

    /* Check column constraints */
    if (column_def_constraints != NIL) {
      ListCell   *l;
      foreach(l, column_def_constraints) {
        Constraint *constr = (Constraint *) lfirst(l);
        if (constr->contype != CONSTR_NOTNULL && constr->contype != CONSTR_PRIMARY) {
          ereport(ERROR,
              (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
                errmsg("Tsurugi supports only NOT NULL and PRIMARY KEY in column constraint")));
          return result;
        }
      }
    }

    /* If COLLATE clause, LIKE clause, DEFAULT constrint, or other syntax, collOid is valid. */
    if ((column_def->collClause != nullptr) | OidIsValid(column_def->collOid)) {
      ereport(ERROR,
              (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
                errmsg("Tsurugi does not support COLLATE clause")));
      return result;
    }

    /* Check for duplicates */
    std::string colname = std::string(column_def->colname);
    if (column_names.find(colname) != column_names.end()) {
      ereport(ERROR,
              (errcode(ERRCODE_DUPLICATE_COLUMN),
                errmsg("column \"%s\" specified more than once",
                      colname.c_str()),
                errdetail("Tsurugi does not support this syntax")));
      return result;
    }
    column_names.insert(colname);
  }

  List *table_constraints = create_stmt->constraints;

  /* Check table constraints */
  foreach(l, table_constraints) {
    Constraint* constr = (Constraint*) lfirst(l);
    if (constr->contype != CONSTR_PRIMARY) {
        ereport(ERROR,
            (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
              errmsg("Tsurugi supports only PRIMARY KEY in table constraint")));
        return result;
    }
  }
#if 0
  /* Check members of IndexStmt structure */
  if (index_stmt != nullptr) {
    if (index_stmt->unique && !(index_stmt->primary)) {
        show_table_constraint_syntax_error_msg("Tsurugi does not support UNIQUE table constraint");
        return result;
    }

    if (index_stmt->tableSpace != nullptr) {
        ereport(ERROR,
                (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
                  errmsg("Tsurugi does not support USING INDEX TABLESPACE clause")));
        return result;
    }

    if (index_stmt->indexIncludingParams != NIL) {
        ereport(ERROR,
                (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
                  errmsg("Tsurugi does not support INCLUDE clause")));
        return result;
    }

    if (index_stmt->options != NIL) {
        ereport(ERROR,
                (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
                  errmsg("Tsurugi does not support WITH clause")));
        return result;
    }

    if (index_stmt->excludeOpNames != nullptr) {
        show_table_constraint_syntax_error_msg("Tsurugi does not support EXCLUDE table constraint");
        return result;
    }

    if (index_stmt->deferrable) {
        ereport(ERROR,
                (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
                  errmsg("Tsurugi does not support DEFERRABLE")));
        return result;
    }

    if (index_stmt->initdeferred) {
        ereport(ERROR,
                (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
                  errmsg("Tsurugi does not support INITIALLY DEFERRED")));
        return result;
    }
  }
#endif
  /**
    * If statements to create temprary or unlogged table are given,
    * report error messages.
    */
  RangeVar* relation = (RangeVar*) create_stmt->relation;
  if (relation != nullptr && relation->relpersistence != RELPERSISTENCE_PERMANENT) {
    ereport(ERROR,
        (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
          errmsg("Tsurugi supports only regular table")));
    return result;
  }

  result = true;

  return result;
}

/**
 *  @brief  Check if given syntax supported or not by Tsurugi
 *  @return true if supported
 *  @return false otherwise.
 */
bool CreateTable::validate_data_type() const
{
  assert(create_stmt() != nullptr);

  bool result{true};
  CreateStmt* create_stmt = this->create_stmt();
  List *table_elts = create_stmt->tableElts;
  ListCell *l;
  auto datatypes = std::make_unique<metadata::DataTypes>("tsurugi");

  /*
    * List of TypeName structure's member "names" not supported by Tsurugi.
    * If type_oid_not_supported is not empty, type_names_not_supported is NIL.
    */
  List *type_names_not_supported = NIL;
  /*
    * List of TypeName structure's member "typeOid" not supported by Tsurugi.
    * If type_oid_not_supported is not NIL, type_oid_not_supported is empty.
    */
  std::vector<int> type_oid_not_supported;

  /* Check type of each column */
  foreach(l, table_elts) {
    ColumnDef *column_def = (ColumnDef *)lfirst(l);
    TypeName *column_def_type_name = column_def->typeName;

    if (column_def_type_name != nullptr) {
      List *type_names = column_def_type_name->names;

      /**
        * If "names" is NIL then the actual type OID is given by typeOid,
        * otherwise typeOid is unused.
        */
      if (type_names != NIL) {
        bool is_supported{false};

        ListCell   *l;
        foreach(l, type_names) {
          Value *type_name_value = (Value *)lfirst(l);
          std::string type_name{std::string(strVal(type_name_value))};

          ptree datatype;
          /*
            * ErrorCode::OK if the given type name is suppoted by Tsurugi,
            * otherwize error code is returned.
            */
          metadata::ErrorCode err = datatypes->get(
              metadata::DataTypes::PG_DATA_TYPE_QUALIFIED_NAME, type_name, datatype);
          if (err == metadata::ErrorCode::OK) {
              is_supported = true;
          }
        }

        /* If the given type name is not suppoted, append the list of type names not supported*/
        if (!is_supported) {
            type_names_not_supported = lappend(type_names_not_supported,type_names);
            result = false;
        }
      }

      if (OidIsValid(column_def_type_name->typeOid)) {
        ptree datatype;
        std::string type_oid_str = std::to_string(column_def_type_name->typeOid);

        /*
          * ErrorCode::OK if the given type is suppoted by Tsurugi,
          * otherwize error code is returned.
          */
        metadata::ErrorCode err = datatypes->get(
              metadata::DataTypes::PG_DATA_TYPE, type_oid_str, datatype);
        /* If the given type is not suppoted, append the list of type oids not supported*/
        if (err != metadata::ErrorCode::OK) {
            type_oid_not_supported.push_back(column_def_type_name->typeOid);
            result = false;
        }
      }
    } else {
      /* If invalid syntax is given, report error messages. */
      show_syntax_error_msg("type name of column is not specified");
      result = false;
      return result;
    }
  }

  /* If given type is not supported, report error messages. */
  if (result != true) {
    if (type_names_not_supported != NIL) {
        show_type_error_msg(type_names_not_supported);
    }
    if (!type_oid_not_supported.empty()) {
        show_type_error_msg(type_oid_not_supported);
    }
  }

  return result;
}

/**
 * @brief 	Generate metadata from query tree.
 * @param	table_tree	
 * @return	true if success.
 */
bool CreateTable::generate_metadata2(manager::metadata::Table& table)
{
	const CreateStmt* create_stmt = this->create_stmt();
	assert(create_stmt != nullptr);

	bool result{false};
	ptree table_tree;

	//
	// table metadata
	//
	// table name
	RangeVar* relation = (RangeVar*) create_stmt->relation;
	if (relation != nullptr && relation->relname != nullptr) {
		table_tree.put(metadata::Tables::NAME, relation->relname);
		table.name = relation->relname;
	} else {
		show_syntax_error_msg("table name is not specified");
		return result;
	}
	// tuples
	table_tree.put(metadata::Tables::TUPLES, 0);
	table.tuples = 0;

	//
	// columns metadata
	//
	ptree columns_node;

	/* for each columns */
	List* table_elts = create_stmt->tableElts;
	ListCell* listptr;
	int64_t ordinal_position = ORDINAL_POSITION_BASE_INDEX;
	foreach(listptr, table_elts) {
		Node* node = (Node *) lfirst(listptr);
		if (IsA(node, ColumnDef)) {
			ColumnDef* column_def = (ColumnDef*) node;
			ptree column_tree;
			metadata::Column column;

			bool success = create_column_metadata(column_def, 
													ordinal_position, 
													column_tree, 
													column);
			if (!success) {
				return result;
			}
			columns_node.push_back(std::make_pair("", column_tree));
			table.columns.emplace_back(column);
			ordinal_position++;
		}
	}
	table_tree.add_child(metadata::Tables::COLUMNS_NODE, columns_node);

	//
	// primary keys
	//
	table.primary_keys = get_primary_key(create_stmt, table.columns);
	if (table.primary_keys.size() > 0) {
		ptree primary_key;
		for (auto ordinal_position : table.primary_keys) {
			primary_key.put<int64_t>("", ordinal_position);
			primary_key.push_back(std::make_pair("", primary_key));
		}
		table_tree.add_child(metadata::Tables::PRIMARY_KEY_NODE, primary_key);
	}

	result = true;

	return result;
}

/**
 * @brief
 * @param column_def [in] column query tree.
 * @param ordinal_position [in] column ordinal position
 * @param column [out] column metadata
 */
bool CreateTable::create_column_metadata(ColumnDef* column_def, 
                                          int64_t ordinal_position, 
                                          ptree& column_tree,
                                          metadata::Column& column)
{
	assert(column_def != nullptr);

	bool result = false;
	auto datatypes = std::make_unique<metadata::DataTypes>("tsurugi");

	/* ordinalPosition  */
	column_tree.put<int64_t>(metadata::Tables::Column::ORDINAL_POSITION, ordinal_position);
	column.ordinal_position = ordinal_position;

	/* column name */
	column_tree.put(metadata::Tables::Column::NAME, column_def->colname);
	column.name = column_def->colname;

	// nullable
	column_tree.put<bool>(metadata::Tables::Column::NULLABLE, !column_def->is_not_null);
	column.nullable = !(column_def->is_not_null);

	//
	// column data type
	//
	TypeName *column_def_type_name = column_def->typeName;
	if (column_def_type_name == nullptr) {
		show_syntax_error_msg("type name of the column is not specified");
		return result;
	}
	List *type_names = column_def_type_name->names;
	boost::optional<metadata::ObjectIdType> data_type_id;

	/*
	* If "names" is NIL then the actual type OID is given by typeOid,
	* otherwise typeOid is unused.
	*/
	ListCell* listptr;
	foreach(listptr, type_names) {
		Value* type_name_value = (Value*) lfirst(listptr);
		std::string type_name{std::string(strVal(type_name_value))};

		// PG data type qualified name
		ptree datatype;
		metadata::ErrorCode error = datatypes->get(
				metadata::DataTypes::PG_DATA_TYPE_QUALIFIED_NAME, type_name, datatype);
		if (error == metadata::ErrorCode::OK) {
			data_type_id = 
					datatype.get_optional<metadata::ObjectIdType>(metadata::DataTypes::ID);
			break;
		}
	}

	// data type ID
	if (OidIsValid(column_def_type_name->typeOid)) {
		ptree datatype;
		std::string type_oid_str = std::to_string(column_def_type_name->typeOid);
		metadata::ErrorCode err = datatypes->get(
			metadata::DataTypes::PG_DATA_TYPE, type_oid_str, datatype);
		if (err == metadata::ErrorCode::OK) {
			data_type_id = datatype.get_optional<metadata::ObjectIdType>(metadata::DataTypes::ID);
		}
	}
	if (!data_type_id) {
		ereport(ERROR,
				(errcode(ERRCODE_INTERNAL_ERROR),
				errmsg("Tsurugi could not found data type ids")));
		return result;
	} 
	metadata::ObjectIdType id = data_type_id.get();
	column_tree.put<metadata::ObjectIdType>(metadata::Tables::Column::DATA_TYPE_ID, id);
	column.data_type_id = id;

	/* put varying metadata if given type is varchar or char */
	switch (static_cast<metadata::DataTypes::DataTypesId>(id)) {
		case metadata:: DataTypes::DataTypesId::VARCHAR: {
			/* put true if given type is varchar */
			column_tree.put<bool>(metadata::Tables::Column::VARYING, true);
			column.varying = true;
			break;
		}
		case metadata::DataTypes::DataTypesId::CHAR: {
			/* put false if given type is char */
			column_tree.put<bool>(metadata::Tables::Column::VARYING, false);
			column.varying = false;
			break;
		}
		default:
			/* other data type */
			column.varying = false;
			break;
	}

	/* put data type lengths metadata if given type is varchar or char */
	switch (static_cast<metadata::DataTypes::DataTypesId>(id)) {
		case metadata::DataTypes::DataTypesId::VARCHAR:
		case metadata::DataTypes::DataTypesId::CHAR: {
			/*
				* if "typmods" is NIL then the actual typmod is expected to
				* be prespecified in typemod, otherwise typemod is unused.
				*/
			List *typmods = column_def_type_name->typmods;
			/* typemod includes varlena header */
			int64_t typemod = (int64_t) column_def_type_name->typemod - VARHDRSZ;
			if (typmods != NIL) {
				/* for data type lengths metadata */
				ptree data_lengths;
				bool success = put_data_lengths(typmods, data_lengths);
				if (!success) {
				return result;
				}
				/* put data type lengths metadata */
				if (!data_lengths.data().empty()) {
					column_tree.add_child(metadata::Tables::Column::DATA_LENGTH, data_lengths);
					column.data_length = 99;
				}
			} else if (typemod >= 0) {
				/* if typmod is -1, typmod is NULL VALUE*/
				/* put a data type length metadata */
				column_tree.put<metadata::ObjectIdType>(metadata::Tables::Column::DATA_LENGTH, typemod);
				column.data_length = typemod;
			}
			break;
		}
		default: {
			/* other data type */
			column.data_length = 0;
			break;
		}
	}  
	result = true;

	return result;
}

bool CreateTable::put_data_lengths(List* typmods, ptree& datalengths)
{
	bool result = false;

	ListCell *l;
	foreach(l, typmods) {
	Node* tm = (Node*) lfirst(l);
	if (IsA(tm, A_Const)) {
		A_Const* ac = (A_Const*) tm;
		if (IsA(&ac->val, Integer)) {
			/*
			* get data type lengths from typmods of TypeName structure.
			* The given data type length must be constant integer value.
			*/
			datalengths.put<metadata::ObjectIdType>("", ac->val.val.ival);
		} else {
			show_syntax_error_msg("can use only integer value in data length");
			return result;
		}
	} else {
		show_syntax_error_msg("can use only constant value in data length");
		return result;
	}
	}
	result = true;

	return result;
}
