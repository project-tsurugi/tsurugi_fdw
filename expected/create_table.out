/* DDL */
CREATE TABLE t1(c1 INTEGER NOT NULL PRIMARY KEY) TABLESPACE tsurugi;
CREATE FOREIGN TABLE t1(c1 INTEGER NOT NULL) SERVER tsurugidb;
CREATE TABLE t2(c1 INTEGER NOT NULL PRIMARY KEY, c2 BIGINT, c3 DOUBLE PRECISION) TABLESPACE tsurugi;
CREATE FOREIGN TABLE t2(c1 INTEGER NOT NULL, c2 BIGINT, c3 DOUBLE PRECISION) SERVER tsurugidb;
/* DML */
SELECT * FROM t1;
 c1 
----
(0 rows)

INSERT INTO t1 (c1) VALUES (1);
SELECT * FROM t1;
 c1 
----
  1
(1 row)

SELECT * FROM t2;
 c1 | c2 | c3 
----+----+----
(0 rows)

INSERT INTO t2 (c1, c2, c3) VALUES (1, 100, 1.1);
INSERT INTO t2 (c1, c2, c3) VALUES (2, 200, 2.2);
SELECT * FROM t2;
 c1 | c2  | c3  
----+-----+-----
  1 | 100 | 1.1
  2 | 200 | 2.2
(2 rows)

SELECT * FROM t1;
 c1 
----
  1
(1 row)

INSERT INTO t1 (c1) VALUES (10);
INSERT INTO t1 (c1) VALUES (100);
SELECT * FROM t1;
 c1  
-----
   1
  10
 100
(3 rows)

SELECT * FROM t2;
 c1 | c2  | c3  
----+-----+-----
  1 | 100 | 1.1
  2 | 200 | 2.2
(2 rows)

UPDATE t2 SET c2 = c2+10;
UPDATE t2 SET c3 = c3+1.1 WHERE c2 = 110;
SELECT * FROM t2;
 c1 | c2  | c3  
----+-----+-----
  1 | 110 | 2.2
  2 | 210 | 2.2
(2 rows)

SELECT * FROM t1;
 c1  
-----
   1
  10
 100
(3 rows)

DELETE FROM t1 WHERE c1 = 10;
SELECT * FROM t1;
 c1  
-----
   1
 100
(2 rows)

/* DDL */
DROP TABLE t1;
DROP TABLE t2;
DROP FOREIGN TABLE t1;
DROP FOREIGN TABLE t2;
